Image tracing

dnfield opened the pull request:
## Description

Add tracing to image cache calls outside of release mode.  Looks like this:

![image](https://user-images.githubusercontent.com/8620741/74362702-06970500-4d7e-11ea-9525-63eb1e20dd02.png)

And can be easily lined up with engine work to see what's going on. The small line dropping off of `listener` is to a cleanup routine that runs at the end of listening, which is too small to be visible at the zoom level of the screenshot (but is there!) :).

/cc @liyuqian @chinmaygarde @alml @ignatz @goderbauer @gaaclarke 

## Tests

I added the following tests:

I'm trying to figure out the right way to test this.  I think we'll need to update a driver test and check that this data is in there to start, but I'm open to suggestions.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

dnfield committed:
start

dnfield committed:
..

dnfield committed:
Trace image cache calls

dnfield committed:
test

dnfield commented:
I'm working on tests for this, need to land a couple other things to enable it first - #50663 and #50666

dnfield committed:
Merge remote-tracking branch 'upstream/master' into image_tracing

dnfield committed:
test, packages, merge

dnfield commented:
I tried to keep the package updates out of this but two more snuck in.  Oh well.

dnfield commented on packages/flutter/pubspec.yaml:
```diff
@@ -23,6 +23,7 @@ dev_dependencies:
   flutter_goldens:
     sdk: flutter
   mockito: 4.1.1
+  vm_service: 2.3.1
```
This is the only pubspec change I actually care about for this PR.

dnfield commented on packages/flutter_tools/pubspec.yaml:
```diff
@@ -11,7 +11,7 @@ dependencies:
   archive: 2.0.11
   args: 1.5.2
-  dwds: 1.0.1
+  dwds: 0.4.0
```
@jonahwilliams what's up with this?  Why did it downgrade dwds?

jonahwilliams commented:
:man_shrugging: 

You could try pinning it in update_packages.dart

bkonyi commented on packages/flutter/test/tracing/image_cache_tracing_test.dart:
```diff
@@ -0,0 +1,95 @@
+    List<TimelineEvent> events, List<Map<String, dynamic>> expected) {
+  for (final TimelineEvent event in events) {
+    for (int index = 0; index < expected.length; index += 1) {
+      if (expected[index]['name'] == event.json['name']) {
```
This would be cleaner if you did:

```dart
for (final e in expected) {
  ...
  if (_mapsEqual(expectedArgs, args)) {
    expected.remove(e);
  }
}
```

dnfield replied:
_mapsEqual assumes it's doing a flat map though, and there are no nested maps.  I guess I could update it to check for that...

dnfield replied:
I'm trying to short circuit the more expensive recursion this way though by keying on the name.  It's faster to check that one property than to recurse the entire structure, which could become large as we add to these tests.

bkonyi commented on packages/flutter/lib/src/painting/image_cache.dart:
```diff
@@ -85,11 +89,18 @@ class ImageCache {
       return;
+    TimelineTask timelineTask;
+    if (!kReleaseMode) {
+       timelineTask = TimelineTask()..start('ImageCache.setMaximumSize', arguments: <String, dynamic>{'value': value});
```
Nit: You might want to consider breaking this into multiple lines (below as well)

dnfield replied:
Done

bkonyi reviewed (approved):
Timeline code and test LGTM

dnfield committed:
fix packages

dnfield commented:
This test either has to be rewritten (the dwds pinned version of vm_service is broken for what the test uses), or we have to get dwds to bump vmservice for this.

I'm separately going to open an internal CL to open up visibility for flutter to use this package.

dnfield committed:
make it work

dnfield commented:
I rewrote the test to not depend on package:vm_service.  Filed https://github.com/dart-lang/webdev/issues/899 to track getting that bumped in dwds and added a TODO.

dnfield committed:
newline

dnfield commented on dev/bots/test.dart:
```diff
@@ -457,6 +457,11 @@ Future<void> _runFrameworkTests() async {
     await _runFlutterTest(path.join(flutterRoot, 'packages', 'flutter_test'), tableData: bigqueryApi?.tabledata);
     await _runFlutterTest(path.join(flutterRoot, 'packages', 'fuchsia_remote_debug_protocol'), tableData: bigqueryApi?.tabledata);
+    await _runFlutterTest(
+      path.join(flutterRoot, 'packages', 'flutter', 'tracing'),
```
@yjbanov - do you have any complaints about putting this in package/flutter?

This depends on dart:io and the io methods of connecting to the vm service.  I'm doing this to avoid writing a driver test, since we're not actually using this for performance just yet.  I have it marked as skip if browser.

dnfield replied:
I've thought the better of this.  Having this test under packages/flutter means a regular person can't just run `flutter test` in that folder without at least one test failing.  Moved this to `dev/tracing` with a README.

dnfield committed:
simplify test

dnfield committed:
fix re-entrancy

dnfield committed:
fix bots

dnfield committed:
move to separate location

dnfield committed:
fix test for new location...

dnfield committed:
typo

goderbauer commented on dev/tracing_tests/README.md:
```diff
@@ -0,0 +1,7 @@
+# Tracing tests
+
+The tests in this folder must be run with `flutter test --enable-vmservice`,
```
Any chance you can check for this condition from within the test and just print this message if the test was run incorrectly?

goderbauer replied:
(Readmes usually get ignored..)

goderbauer replied:
Never mind, I see below that you already do this: 🌟

goderbauer commented on dev/tracing_tests/test/image_cache_tracing_test.dart:
```diff
@@ -0,0 +1,161 @@
+import 'dart:isolate' as isolate;
+
+import 'package:flutter/painting.dart';
+
```
nit: I believe this blank line shouldn't be here?

dnfield replied:
Removed.

goderbauer reviewed (commented):
LGTM

dnfield committed:
Update image_cache_tracing_test.dart

jonahwilliams commented on packages/flutter/test/tracing/README.md:
```diff
@@ -0,0 +1,7 @@
+# Tracing tests
```
Should this be removed?

dnfield replied:
The whole file or this line?

I think most of the folders in `dev/` have a README.md - the goal of this one is to make it clear what kind of tests belong in here, and what's special about running them.

jonahwilliams replied:
This isn't in dev though, this is in `flutter/test`

dnfield replied:
Ah! Thanks, good catch.  Removed.

dnfield committed:
oops

jonahwilliams reviewed (approved):
LGTM
